File: libraries/authz-spring-boot/src/main/java/com/example/authz/engine/Scoring.java
package com.example.authz.engine;

import com.example.authz.config.CompiledRule;
import com.example.authz.config.Segment;

import java.util.ArrayList;
import java.util.List;

/** Path matching + specificity scoring (must match the NestJS implementation). */
public final class Scoring {
    private Scoring() {}

    /** Split a request path into segments, ignoring empty parts. */
    public static List<String> splitPath(String path) {
        List<String> out = new ArrayList<>();
        for (String part : path.split("/")) {
            if (!part.isEmpty()) out.add(part);
        }
        return out;
    }

    public static int scoreSegment(Segment seg) {
        return switch (seg.kind()) {
            case LITERAL -> 2;
            case SINGLE -> 1;
            case DEEP -> 0;
        };
    }

    /** `*` matches one segment; `**` matches one or more (final segment only). */
    public static boolean matchPath(List<Segment> segments, List<String> pathSegs) {
        for (int i = 0; i < segments.size(); i++) {
            Segment seg = segments.get(i);
            if (seg.kind() == com.example.authz.config.SegmentKind.DEEP) {
                return pathSegs.size() - i >= 1;
            }
            if (i >= pathSegs.size()) return false;
            if (seg.kind() == com.example.authz.config.SegmentKind.LITERAL
                    && !seg.value().equals(pathSegs.get(i))) {
                return false;
            }
        }
        return segments.size() == pathSegs.size();
    }

    /**
     * &gt;0 if a is more specific than b. Order: segment scores left-to-right,
     * then more literal segments, then longer pattern.
     */
    public static int compareSpecificity(CompiledRule a, CompiledRule b) {
        int n = Math.min(a.scores().length, b.scores().length);
        for (int i = 0; i < n; i++) {
            if (a.scores()[i] != b.scores()[i]) return a.scores()[i] - b.scores()[i];
        }
        if (a.literalCount() != b.literalCount()) return a.literalCount() - b.literalCount();
        if (a.segments().size() != b.segments().size()) {
            return a.segments().size() - b.segments().size();
        }
        return 0;
    }
}

File: libraries/authz-nestjs/src/decision-engine/scoring.ts
import { CompiledRule, Segment } from "../rule-config/types";

/** Split a request path into segments, ignoring leading/trailing slashes. */
export function splitPath(path: string): string[] {
  return path.split("/").filter((s) => s.length > 0);
}

/** Score a single segment: literal=2, single (*) =1, deep (**) =0. */
export function scoreSegment(seg: Segment): number {
  switch (seg.kind) {
    case "literal":
      return 2;
    case "single":
      return 1;
    case "deep":
      return 0;
  }
}

/**
 * Does a compiled pattern match the given path segments?
 * `*` matches exactly one segment; `**` matches one or more (final segment only).
 */
export function matchPath(segments: Segment[], pathSegs: string[]): boolean {
  for (let i = 0; i < segments.length; i++) {
    const seg = segments[i];
    if (seg.kind === "deep") {
      // ** is always the final segment; it must consume at least one segment.
      return pathSegs.length - i >= 1;
    }
    if (i >= pathSegs.length) return false;
    if (seg.kind === "literal" && seg.value !== pathSegs[i]) return false;
    // single matches any one segment
  }
  return segments.length === pathSegs.length;
}

/**
 * Compare specificity of two compiled rules.
 * Returns >0 if `a` is more specific (should be preferred), <0 if `b` is, 0 if tied.
 * Order: segment scores left-to-right, then more literal segments, then longer pattern.
 */
export function compareSpecificity(a: CompiledRule, b: CompiledRule): number {
  const n = Math.min(a.scores.length, b.scores.length);
  for (let i = 0; i < n; i++) {
    if (a.scores[i] !== b.scores[i]) return a.scores[i] - b.scores[i];
  }
  if (a.literalCount !== b.literalCount) return a.literalCount - b.literalCount;
  if (a.segments.length !== b.segments.length) {
    return a.segments.length - b.segments.length;
  }
  return 0;
}
